### PR TITLE
CRF for keras 2.x

### PR DESCRIPTION
**Pull request by linxihui**
An Implementation of linear chain conditional random field (CRF), modified for keras 2.0 from previous PR https://github.com/fchollet/keras/pull/4646

Note this PR has https://github.com/farizrahman4u/keras-contrib/pull/79 merged. Please merged.

### CODE REVIEW

**Comment by linxihui** [comment]
@farizrahman4u The tensorflow and keras version are not compatible in .travis checking.
The tensorflow version is 0.12. It has two concatenate functions `tf.concat(axis, tensors)` and `tf.concat_v2(tensors, axis)`. However, in keras tensorflow backend, `K.concatenate(tensors, axis) = tf.concat(tensors, axis)`. In tensorflow 1.x, `tf.concat` is changed to accept `tensors` as 1st argument.  This is the reason the CRF checking fail.

I think tensorflow should be update to version 1.x in .travis.

**Comment by farizrahman4u** [comment]
@linxihui Thanks for pointing it out.. please submit a PR to fix travis config so that you get the Github credits.

**Comment by farizrahman4u** [comment]
Cool stuff! Can anyone review?

File: keras_contrib/layers/crf.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+        def acc(y_true, y_pred):
+            X = self.inbound_nodes[0].input_tensors[0]
+            mask = self.inbound_nodes[0].input_masks[0]
+            y_pred = self.get_marginal_prob(X, mask)
+            return self._get_accuracy(y_true, y_pred, mask, self.sparse_target)
+        acc.func_name = 'marginal_acc'
+        return acc
+
+    @staticmethod
+    def log_sum_exp(x, axis=-1):
```
Comment (tboquet) [inline keras_contrib/layers/crf.py]: Since fchollet/keras#6346 has been merged, could it be integrated in this PR?

File: examples/conll2000_chunking_crf.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+"""Train CRF and BiLSTM-CRF on CONLL2000 chunking data, similar to https://arxiv.org/pdf/1508.01991v1.pdf.
+"""
+from __future__ import absolute_import
+from __future__ import print_function
+
+import numpy
+from collections import Counter
+from sklearn.metrics import classification_report
```
Comment (tboquet) [inline examples/conll2000_chunking_crf.py]: Should we try to avoid those dependencies (`sklearn` and `nltk`)?
I agree this is very convenient and not hard to install but it could be easy to add a new dataset as in #25.

File: keras_contrib/layers/crf.py (hunk 1 above)
Comment (linxihui) [inline keras_contrib/layers/crf.py]: Sure, will change it soon.

File: examples/conll2000_chunking_crf.py (hunk 2 above)
Comment (linxihui) [inline examples/conll2000_chunking_crf.py]: `sklearn` is not necessary, as I just use the `classification_report` to print the result in a nice way. `nltk` is here for getting the data. My  original idea was a bit on the opposite, avoiding having the same dataset in multiple packages or writing duplicate script to load data, but instead using an existing package, which also saves the effort of licensing.  But I understand the disadvantage as well. I'll change it, as it probably fits better to the Keras way.

File: keras_contrib/layers/crf.py (hunk 1 above)
Comment (tboquet) [inline keras_contrib/layers/crf.py]: Ok let's keep this in mind and open an issue when this will be merged?

Comment (linxihui) [inline keras_contrib/layers/crf.py]: Then I probably should just change it.  Actually, there is a bug when using `K.logsumexp` with` `tensorflow` backend, when passed a placeholder, it losses the shape information, i.e., `K.int_shape` got None. This is due to the fact that `K.logsumexp` uses an deprecidated argument. I'll make a PR to `keras` and will make the change here one it is merged.

**Comment by linxihui** [comment]
@tboquet  using `K.logsumexp` now.

**Comment by linxihui** [comment]
@tboquet  @farizrahman4u  any one do the code review?

**Comment by farizrahman4u** [comment]
LGTM, need confirmation from one more moderator.
